pkg/xds/context/mesh_context_builder.go

```diff
@@ -156,19 +241,29 @@ func (m *meshContextBuilder) fetchResources(ctx context.Context, mesh *core_mesh
 
 			insights := &core_mesh.ServiceInsightResourceList{}
 			if err := m.rm.List(ctx, insights, core_store.ListByMesh(mesh.Meta.GetName())); err != nil {
-				return nil, err
+				return Resources{}, err
 			}
 
-			resources[typ] = insights
+			resources.MeshLocalResources[typ] = insights
+		case core_mesh.MeshGatewayType:
+			if err := m.fetchCrossMesh(ctx, typ, meshName, &resources, nil); err != nil {
+				return Resources{}, err
+			}
+		case core_mesh.DataplaneType:
+			if err := m.fetchCrossMesh(ctx, typ, meshName, &resources, func(dp core_model.Resource) bool {
+				return dp.(*core_mesh.DataplaneResource).Spec.IsBuiltinGateway()
+			}); err != nil {
+				return Resources{}, err
+			}
```

Comment:
I'm not sure if this is really a problem or not.

MeshContextBuilder uses cached ReadOnlyResourceManager.
Technically only one Mesh needs to retrieve all DPs, other will most likely pick it up.

What I think may help is to pass otherMeshes to `fetchCrossMesh` and do `ListByMesh`. This way we can leverage existent cached queries of `fetch all dps from mesh X` that are executed anyways for all the meshes in parallel.

And yes, ideally we would not have this dependency. We could also
1) Build one intermediate Config object for each mesh that is aggregation of all needed information from all Gateway instances and have dependency on that.
2) Create a separate Gateway object (instead of Dataplane), just like we did with ZoneIngress/ZoneEgress

Reply:
Yes Cache would help but all caches have limits and you still end up iterating over things. As I said it's an optimisation anyway...
I think 1. and 2. would both make sense. A gateway is different enough that it probably doesn't make sense to just shove this in the DPP resource. I'm going to make sure we have issues about these (also adding an issue for moving ZoneIngress services outside of the ZoneIngress resource and have them be per mesh).

Reply:
1. #4359
2. #4358
3. (ZoneIngress split) #4360